openedx_certificates/models.py

```diff
@@ -1,5 +1,345 @@
+
+        :param user_ids: A list of user IDs to filter.
+        :return: A list of user IDs that either:
+                 1. Do not have a certificate for this course and certificate type.
+                 2. Have such a certificate with an error status.
+        """
+        # TODO: Delete this after testing.
+        return user_ids
+        users_ids_with_certificates = ExternalCertificate.objects.filter(
+            models.Q(course_id=self.course_id),
+            models.Q(certificate_type=self.certificate_type),
+            ~(models.Q(status=ExternalCertificate.Status.ERROR)),
+        ).values_list('user_id', flat=True)
+
+        filtered_user_ids_set = set(user_ids) - set(users_ids_with_certificates)
+        return list(filtered_user_ids_set)
+
+    def get_eligible_user_ids(self) -> list[int]:
+        """
+        Get the list of eligible learners for the given course.
+
+        :return: A list of user IDs.
+        """
+        func_path = self.certificate_type.retrieval_func
+        module_path, func_name = func_path.rsplit('.', 1)
+        module = import_module(module_path)
+        func = getattr(module, func_name)
+
+        custom_options = {**self.certificate_type.custom_options, **self.custom_options}
+        return func(self.course_id, custom_options)
+
+    def generate_certificate_for_user(self, user_id: int, celery_task_id: int = 0):
+        """
+        Celery task for processing a single user's certificate.
+
+        This function retrieves an ExternalCertificateCourse object based on course_id and certificate_type_id,
+        retrieves the data using the retrieval_func specified in the associated ExternalCertificateType object,
+        and passes this data to the function specified in the generation_func field.
+
+        Args:
+            user_id: The ID of the user to process the certificate for.
+            celery_task_id (optional): The ID of the Celery task that is running this function.
+        """
+        user = get_user_model().objects.get(id=user_id)
+        # Use the name from the profile if it is not empty. Otherwise, use the first and last name.
+        # We check if the profile exists because it is absent in unit tests.
+        user_full_name = getattr(getattr(user, 'profile', None), 'name', f"{user.first_name} {user.last_name}")
+        custom_options = {**self.certificate_type.custom_options, **self.custom_options}
+
+        try:
+            certificate = ExternalCertificate.objects.get(
+                user_id=user_id,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+            )
+            certificate.user_full_name = user_full_name
+            certificate.status = ExternalCertificate.Status.GENERATING
+            certificate.generation_task_id = celery_task_id
+            certificate.save()
+        except ExternalCertificate.DoesNotExist:
```

Comment:
Nit: Seems like you could use [`update_or_create`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#update-or-create) here to shorten this code.